theano/tensor/opt.py

```diff
@@ -4246,6 +4330,43 @@ def local_sum_alloc(node):
 
 
 @register_specialize
+@gof.local_optimizer([T.elemwise.Prod])
+def local_prod_alloc(node):
+    """ prod(alloc(constant,shapes...)) => constant**prod(shapes)"""
+    if isinstance(node.op, T.elemwise.Prod):
+        prod_inps, = node.inputs
+        if prod_inps.owner and isinstance(summed.owner.op, T.Alloc):
+            input = prod_inps.owner.inputs[0]
+            shapes = prod_inps.owner.inputs[1:]
+
+            # ask arnaud
+            # what is this checking for
+            if (node.op.axis is None or
+                node.op.axis == tuple(range(input.ndim))):
+                try:
+                    val = get_scalar_constant_value(input)
+                    assert val.size == 1
+                    val = val.reshape(1)[0] * T.mul(*shapes)
+                    return [T.cast(val, dtype=node.outputs[0].dtype)]
+                except NotScalarConstantError:
+                    pass
+            else:
+                try:
+                    val = get_scalar_constant_value(input)
+                    assert val.size == 1
+                    val = val.reshape(1)[0]
+                    to_prod = [shapes[i] for i in xrange(len(shapes))
+                               if i in node.op.axis]
+                    if to_prod:
+                        val *= T.mul(*to_prod)
```

Comment:
This is wrong.  In the case of prod, we need to do the exponent, not the multiplication.